dask_searchcv/adaptive.py

```diff
@@ -0,0 +1,447 @@
+import math
+import numpy as np
+import scipy.stats as stats
+from functools import partial
+from toolz import reduce
+import toolz
+import warnings
+from timeit import default_timer
+import logging
+import distributed.joblib
+import joblib
+from distutils.version import LooseVersion
+
+import sklearn
+from dask_ml.model_selection._split import train_test_split
+from dask_ml.wrappers import ParallelPostFit
+
+from sklearn import clone
+from sklearn.model_selection import ParameterSampler
+from dask_ml.wrappers import Incremental
+
+import dask.array as da
+from .model_selection import DaskBaseSearchCV, GridSearchCV, _RETURN_TRAIN_SCORE_DEFAULT
+
+logger = logging.getLogger(__name__)
+
+
+def _train(model, X, y, X_val, y_val, max_iter=1, dry_run=False,
+           s=1, i=1, k='_', **fit_kwargs):
+    """
+    Train function. Returns validation score.
+
+    Parameters
+    ----------
+    model : sklearn model to train
+    X, y, X_val, y_val : dask arrays that are training and validation sets
+    max_iter : float, number of times to call partial_fit
+    dry_run : whether to call partial_fit and score or not
+    s, i, k : int or string, used for logging
+    fit_kwargs : dict, kwargs to pass to partial_fit
+
+    Takes max_iters as parameters. THese are treated as a "scarce resource" --
+    Hyperband is trying to minimize the total iters use for max_iters.
+
+    Returns
+    -------
+    val_score : float. Validation score
+    times : dict. keys of ``fit_time`` and ``score_time``, values in seconds.
+
+    """
+    start_time = default_timer()
+
+    if hasattr(X, 'result'):
+        X = X.result()
+    if hasattr(y, 'result'):
+        y = y.result()
+    if hasattr(X_val, 'result'):
+        X_val = X_val.result()
+    if hasattr(y_val, 'result'):
+        y_val = y_val.result()
+
+    for iter in range(int(max_iter)):
+        msg = ("Training model {k} in bracket s={s} iteration {i}. "
+               "This model is {percent:.1f}% trained for this bracket")
+        logger.info(msg.format(iter=iter, k=k, s=s, i=i,
+                               percent=iter * 100.0 / max_iter))
+        if not dry_run:
+            model.partial_fit(X, y, **fit_kwargs)
+    fit_time = default_timer() - start_time
+    msg = ("Training model {k} for {max_iter} partial_fit calls took {secs} seconds")
+    logger.info(msg.format(k=k, max_iter=max_iter, secs=fit_time))
+
+    start_time = default_timer()
+    #  score = model.score(X_val, y_val) if not dry_run else np.random.rand()
+    score = model.score(X_val, y_val) if not dry_run else np.random.rand()
```

Comment:
Extra comment here.

Even though this is an internal API, returning a random value for dry-run here seems fragile. Can you return `None` instead? If not, we'll want to avoid affecting the global random state with calls to `np.random`. Perhaps just `-1` or `np.inf` would work.

Additionally